mlir/include/mlir/Dialect/Mesh/IR/MeshOps.td

```diff
@@ -129,15 +328,28 @@ def Mesh_ShardOp : Mesh_Op<"shard", [
     as an operand in subsequent operations. If not, the sharding applies to the
     operation that defines the tensor value.
 
+    4. `force`: A unit attribute requesting an explicit sharding of the data,
+    therefore not allowing to be optimizied away. This is useful in the presence
+    of halos and inplace semantics.
```

Comment:
That's a possible use. I was primarily focused on inplace semantics like a mutating `insert_slice` which requires a halo-update after the write. I know this goes beyond the current tensor semantics, but we are looking into arrays which do have mutable semantics. An alternative for this case would be an extension of the `ShardingInterface` to indicate which operand a given operation it mutates.

Reply:
Could you add explanation in the doc with an example?
I still don't understand. It would be desirable to avoid the complexity of another attribute.
Shouldn't `insert_slice` know how to handle its spmdization? It would have requirements that the destination-passing style operand and the result are related and that they need to have the same sharding. If this constraint can not be satisfied then resharding will be inserted during sharding propagation.

Reply:
This might well be a lack of understanding on my side.

An `insert_slice` op does not have a result. So I am not sure I fully follow your line of thought.

This was not meant for the sharding propagation, I agree that this does not need this attribute. 

It was meant to serve the spmdization: when spmdization converts `shard` ops into communication ops, e.g. when the input/operand of a `shard` op is another `shard`. If the shardings of the `shard` ops are identical, the algorithm will simply do nothing, e.g. no sharding/communication will happen. However, an in-place operation will require a halo-update even if the sharding stays identical all the time. The elimination of the `shard` op would be incorrect.

```
...
%x = ...
%t = ...
%sharding = mesh.sharding halo_sizes = [2,2] ...
%t1 = mesh.shard %t %sharding
%x1 = mesh.shard %x %sharding
insert_slice %t1 %x1
%t2 = mesh.shard %t1  // <- this must remain and convert to a halo update
%t3 = mesh.shard %t2
<use %t3>
...
```

One solution could be to let the propagation pass insert an explicit `update_halo` op followed by a `shard` op. This would be pretty specific to halos (and not a s generic as a forced resharding) but maybe that is good enough for now. It might also make passes more complicated because now sharding propagation will create IR with sharding ops as well as explicit communication.

Not sure which one is better. Other suggestions?

Reply:
Oh, maybe you are suggesting to handle this in the `spmdize` for `insert_slice`. Not sure if how that could handle a change in halo-size.

Reply:
This conversation gets hidden by github, so I am mentioning @sogartar just in case...

Reply:
`tensor.insert_slice` has a result. It is also a `Pure` operation. Not using its result when this is the intention is an error. Dead code elimination would remove it. In your example, if we are using the result of `tensor.insert_slice`, canonicalization or some other simplification pass should drop the chained `mesh.shard` ops as they refer to the same sharding.
```
...
%x = ...
%t = ...
%sharding = mesh.sharding halo_sizes = [2,2] ...
%t1 = mesh.shard %t %sharding
%x1 = mesh.shard %x %sharding
%t2 = insert_slice %t1 %x1
%t3 = mesh.shard %t2 %sharding
<use %t3>
...
```

Reply:
Yes, `tensor.insert_slice` does have a return argument. Handling that is relatively straight forward. 
However, array semantics (like numpy) do have inplace semantics (tensors do not!). I was talking about such an inplace `insert_slice` operation which is guaranteed to reuse the existing allocation. The tensor dialect does not have such a thing because it is value-based.
I hope this will become more obvious with my next PR.